File: src/main/java/com/itstudio/bookservice/book_service/ports/BookService.java
package com.itstudio.bookservice.book_service.ports;

import com.itstudio.bookservice.book_service.domain.Book;

import java.util.List;
import java.util.Optional;

public interface BookService {
  List<Book> getAll();
  Optional<Book> getById(Long id);
  Book create(Book book);
  Book update(Long id, Book book);
  void delete(Long id);
}

File: src/main/webapp/app/services/book.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Book } from './Book';

@Injectable({
  providedIn: 'root',
})
export class BookService {
  private baseUrl = 'http://localhost:8080/api/books';

  constructor(private http: HttpClient) {}

  getAll(): Observable<Book[]> {
    return this.http.get<Book[]>(this.baseUrl);
  }

  get(id: number): Observable<Book> {
    return this.http.get<Book>(`${this.baseUrl}/${id}`);
  }

  create(book: Book): Observable<Book> {
    return this.http.post<Book>(this.baseUrl, book);
  }

  update(id: number, book: Book): Observable<Book> {
    return this.http.put<Book>(`${this.baseUrl}/${id}`, book);
  }

  delete(id: number): Observable<any> {
    return this.http.delete(`${this.baseUrl}/${id}`);
  }
}
